app/controllers/pipeline_work_in_progress_controller.rb

```diff
@@ -0,0 +1,144 @@
+
+    combined_relationships = {}
+    pipeline_configs.each do |pc|
+      pc.relationships.each do |key, value|
+        combined_relationships[key] ||= []
+        combined_relationships[key] << value
+      end
+    end
+
+    all_purposes = (combined_relationships.keys + combined_relationships.values.flatten).uniq
+
+    # Any purposes with no 'child' purposes should go at the end of the list
+    without_child = all_purposes.select { |p| !(combined_relationships.key? p) }
+
+    while combined_relationships.size > 0
+      # Find any purposes with no 'parent' purposes - to go on the front of the list
+      with_parent = combined_relationships.values.flatten.uniq
+      without_parent = all_purposes - with_parent
+      raise "Pipeline config can't be flattened into a list of purposes" unless without_parent # important to prevent infinite looping
+
+      ordered_purpose_list += without_parent
+
+      # Delete the nodes that have been added, making the next set of purposes have no parent
+      # So we can use the same technique again in the next iteration
+      without_parent.each { |n| combined_relationships.delete(n) }
+
+      # Refresh the all_purposes list for the next iteration
+      all_purposes = (combined_relationships.keys + combined_relationships.values.flatten).uniq
+    end
+
+    # When we've run out of 'parent' purposes, add the final ones on the end
+    ordered_purpose_list += without_child
+    ordered_purpose_list
+  end
+
+
+  def retrieve_labware(page_size, from_date, purposes)
+    p = Sequencescape::Api::V2::Labware
+      .select(
+        {plates: ["uuid", "purpose", "labware_barcode", "state_changes", "created_at"]},
+        {tubes: ["uuid", "purpose", "labware_barcode", "state_changes", "created_at"]},
+        {purposes: "name"}
+      )
+      .includes(:state_changes)
+      .includes(:purpose)
+      .where(
+                without_children: true,
+        purpose_name: purposes,
+        created_at_gt: from_date
+      )
+      .order(:created_at)
+      .per(page_size)
+
+    all_records = []
+    page_num = 1
+    num_retrieved = page_size
+    while num_retrieved == page_size
+      puts "page_num: #{page_num}"
+
+      num_retrieved = p.page(page_num).to_a.size
+      puts "num_retrieved: #{num_retrieved}"
+
+      all_records += p.page(page_num).to_a
+      page_num += 1
+    end
+
+    all_records
+  end
+
+
+  # Returns following structure (example):
+  #
+  # {
+  #   "LTHR Cherrypick" => [
+  #     {
+  #       :record => #<Sequencescape::Api::V2::Labware:@attributes={"type"=>"plates", "id"=>"1234", "uuid"=>"12a34b5c-defg-67hi-8jkl-mn901opq2345", "labware_barcode"=>#<LabwareBarcode:0x00007ff16e5f06b0 @human="DN123456D", @machine="DN123456D", @ean13="1234567890123">, "created_at"=>2021-05-04 12:19:32 +0100}>,
+  #       :state => "pending"
+  #     }
+  #   ],
+  #   "LTHR-384 PCR 1" => [{}]
+  # }
+  def mould_data_for_view(purposes, labware_records)
+    output = {}
```

Comment:
You can use tap here
`{}.tap do |output|
...
end`

It returns the hash.

Reply:
looks good, thanks